contracts/multisig/src/signing.rs

```diff
@@ -29,47 +29,64 @@ impl SigningSession {
         }
     }
 
+    /// returns true if session state was changed
     pub fn recalculate_session_state(
         &mut self,
         signatures: &HashMap<String, Signature>,
         snapshot: &Snapshot,
         block_height: u64,
-    ) {
+    ) -> bool {
         let weight = signers_weight(signatures, snapshot);
 
-        if self.state == MultisigState::Pending && weight >= snapshot.quorum.into() {
-            self.state = MultisigState::Completed {
-                completed_at: block_height,
-            };
+        match self.state {
+            MultisigState::Pending if weight >= snapshot.quorum.into() => {
+                self.state = MultisigState::Completed {
+                    completed_at: block_height,
+                };
+
+                true
+            }
+            _ => false,
         }
     }
 }
 
 pub fn validate_session_signature(
     session: &SigningSession,
-    key: &Key,
     signer: &Addr,
     signature: &Signature,
+    pub_key: &PublicKey,
+    grace_period: u64,
+    block_height: u64,
 ) -> Result<(), ContractError> {
-    // TODO: revisit again once expiration and/or rewards are introduced
-    if matches!(session.state, MultisigState::Completed { .. }) {
+    if matches!(session.state, MultisigState::Completed { completed_at } if completed_at + grace_period < block_height)
+    {
         return Err(ContractError::SigningSessionClosed {
             session_id: session.id,
         });
     }
 
+    if !signature.verify(&session.msg, pub_key)? {
+        return Err(ContractError::InvalidSignature {
+            session_id: session.id,
+            signer: signer.into(),
+        });
+    }
+
```

Comment:
I feel like this function should check whether the signature is a duplicate for this session. I know that check is being done elsewhere, but I feel like it might be cleaner to have all of the validity checks here.

Reply:
this was my first approach, but I don't know how to do it without it looking even weirder, because to do that check here I'll need to first load outside the `Option<Signature>` and pass it here as a parameter just to check if it contains something. For me that looks weirder, thoughts?

Reply:
Well you could just load the signature here, and then store the signature later on, instead of doing an update. It's up to you though, I can see the benefit of having the storage access occur in one place. I just generally view the duplicate signature check as a validation check, and it would be nice to have all of the validation checks in one place. It would also be nice to have the check for whether the signer is a participant in here too. Then you have one function that does all of the checks.